test/SILOptimizer/specialize_deep_generics.swift

```diff
@@ -0,0 +1,34 @@
+// RUN: %target-swift-frontend -O -emit-sil %s | FileCheck %s
+
+// Check that this example does not hang a compiler in the 
+// generic specializer.
```

Comment:
Why not just say "Church Numerals" instead of example. Then it is 100% clear without reading the code what this test case does.